luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,216 @@
+	if (AimingState != AIMING_PREFERRED){
+		return(0);
+	}
+5. in the deferred AimWeaponX() function, add the following at the beginning:
+	if (AimingState != AIMING_DEFERRED){
+		return(0);
+	}
+6. If using a dummy weapon, return (0); in its AimWeaponX() function and QueryWeaponX(piecenum) should be set to static piece lower than the turret.
+
+****OPTIONAL*****
+use weapondef.customparams.smart_error_frames to override the default reloadtime derivitive frames error threshold
+]]
+
+--static
+local frameCheckModulo = Game.gameSpeed
+local failedToFireMultiplier = Game.gameSpeed * 1.25
+local minimumFailedToFireFrames = Game.gameSpeed * 4
+local aggroDecayRate = 0.65
+local aggroDecayCap = 10
+local tallyDecayRate = 0.99
+local pAutoAggro = 12
+local pManualAggro = pAutoAggro * 4
+local dAutoAggro = 4
+local dManualAggro = pAutoAggro * 3
+local dErrorAggro = -300
+local errorMultiplierAddition = 1
+local switchDeferredThreshold = -dAutoAggro * 1.5
+local switchPreferredThreshold = -1
+local AimingState_Preferred = 1
+local AimingState_Deferred = 2
+
+--variables
+local gameFrame = 0
+
+--functions
+local spCallCOBScript = Spring.CallCOBScript
+local spGetUnitWeaponHaveFreeLineOfFire = Spring.GetUnitWeaponHaveFreeLineOfFire
+local spGetUnitWeaponTarget = Spring.GetUnitWeaponTarget
+local spGetUnitWeaponState = Spring.GetUnitWeaponState
+local spGetUnitEstimatedPath = Spring.GetUnitEstimatedPath
+local mathMin = math.min
+local mathMax = math.max
+
+--tables
+local smartUnits = {}
+local unitDefData = {}
+
+function gadget:Initialize()
+	local units = Spring.GetAllUnits()
+	local spGetUnitDefID = Spring.GetUnitDefID
+	for i = 1, #units do
+		gadget:UnitCreated(units[i], spGetUnitDefID(units[i]))
+	end
+end
+
+for unitDefID, def in ipairs(UnitDefs) do
+	if def.weapons then
+		local weapons = def.weapons
+		for weaponNumber, weaponData in pairs(weapons) do
+			local weaponDefID = weapons[weaponNumber].weaponDef
+			if WeaponDefs[weaponDefID] and WeaponDefs[weaponDefID].customParams then
+				if WeaponDefs[weaponDefID].customParams.smart_preferred_weapon then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].preferredWeapon = weaponNumber
+					unitDefData[unitDefID].failedToFireFrameThreshold = WeaponDefs[weaponDefID].customParams
+						.smart_error_frames or
+						mathMax(WeaponDefs[weaponDefID].reload * failedToFireMultiplier, minimumFailedToFireFrames)
+					if def.speed and def.speed ~= 0 then
+						unitDefData[unitDefID].canMove = true
+					end
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_deferred_weapon then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].deferredWeapon = weaponNumber
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_trajectory_checker then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].trajectoryCheckWeapon = weaponNumber
+				end
```

Comment:
How urgent is this redesign? Ideally it would wait until the engine fixes the inability to use arbitrary coordinate for HaveFreeLineOfFire checks, removing the need for a dummy trajectory checker weapon.

Reply:
Personally I think the problem is underestimated and will take significantly longer than you guys anticipate. I commented on that particular issue my thoughts and struggles getting this to work.